staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -796,6 +797,14 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	recursive := opts.Recursive
 	resourceVersion := opts.ResourceVersion
 	pred := opts.Predicate
+	// For recursive lists, we need to make sure the key ended with "/" so that we only
+	// get children "directories". e.g. if we have key "/a", "/a/b", "/ab", getting keys
+	// with prefix "/a" will return all three, while with prefix "/a/" will return only
+	// "/a/b" which is the correct answer.
+	if opts.Recursive && !strings.HasSuffix(key, "/") {
+		key += "/"
```

Comment:
I would not change the key here because the request can be directed to etcd.
I would change the key only if the request would be handled by the cacher.

Reply:
etcd would do the same anyway:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L658-L665

so we can do that unconditionally

Reply:
Doesn't matter, it's an idempotent operation.

Reply:
Ok, found one side effect. With adding suffix before calling Storage, we cause the change in key reported by tracing and possible other places. 

cc @wojtek-t 
on what would be the best way forward.

Reply:
let's do the same thing we do in etcd

preparedKey := key

and then use preparedKey in below places that are not calling underlying storage.

Reply:
Done